### PR TITLE
WIP - Crypto

### PR DESCRIPTION
**Pull request by jsom**

### CODE REVIEW

File: TradeItIosTicketSDK2/EquitySymbolDataSource.swift [hunk 1]
```diff
@@ -0,0 +1,121 @@
+import PromiseKit
+
+protocol SymbolDataSource: UITableViewDataSource {
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/EquitySymbolDataSource.swift]: Seems we should rename this file to be `SymbolDataSource`

File: TradeItIosTicketSDK2/EquitySymbolDataSource.swift [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            let topResults = symbols.filter { $0.symbol?.contains(queryText) ?? false }.prefix(20)
+            self.results = Array(topResults)
+            onSuccess()
+        }.catch { error in
+            print("ERROR") // TODO: Handle better
+            onFailure()
+        }
+    }
+
+    // TODO: Move to getter?
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/EquitySymbolDataSource.swift]: Yea why not -- it's built into the language so I'd prefer this pattern `¯\_(ツ)_/¯`

File: TradeItIosTicketSDK2/EquitySymbolDataSource.swift [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        onFailure: @escaping () -> Void
+    ) {
+        let queryText = textField.text ?? ""
+
+        getAllSymbolsPromise().then { (symbols) -> Void in
+            let topResults = symbols.filter { $0.symbol?.contains(queryText) ?? false }.prefix(20)
+            self.results = Array(topResults)
+            onSuccess()
+        }.catch { error in
+            print("ERROR") // TODO: Handle better
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/EquitySymbolDataSource.swift]: Are we waiting to address this?

File: TradeItIosTicketSDK2/EquitySymbolDataSource.swift [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            onFailure()
+        }
+    }
+
+    // TODO: Move to getter?
+    private func getAllSymbolsPromise() -> Promise<[TradeItSymbol]> {
+        if let resultsPromise = self.resultsPromise { return resultsPromise }
+
+        let resultsPromise = Promise<[TradeItSymbol]> { fulfill, reject in TradeItSDK.symbolService.cryptoSymbols(
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/EquitySymbolDataSource.swift]: new line after `in`?

File: TradeItIosTicketSDK2/EquitySymbolDataSource.swift [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    func searchByTextField(
+        textField: UITextField,
+        onSuccess: @escaping () -> Void,
+        onFailure: @escaping () -> Void
+    ) {
+        let queryText = textField.text ?? ""
+
+        getAllSymbolsPromise().then { (symbols) -> Void in
+            let topResults = symbols.filter { $0.symbol?.contains(queryText) ?? false }.prefix(20)
+            self.results = Array(topResults)
```
Comment (oliver-was-here) [inline TradeItIosTicketSDK2/EquitySymbolDataSource.swift]: Is `topResults` not already `[TradeItSymbol]`

File: TradeItIosEmsApi/TradeItSymbol.h [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#import <Foundation/Foundation.h>
+#import <JSONModel/JSONModel.h>
+
+@protocol TradeItSymbol
+@end
+
+@interface TradeItSymbol : JSONModel<NSCopying>
+
+// The company street symbol
+@property (nullable, copy) NSString<Optional> *symbol;
```
[the remark this answers is not in the document]
Comment (jsom) [reply TradeItIosEmsApi/TradeItSymbol.h]: Should be nonatomic. This was a moved file.

File: TradeItIosEmsApi/TradeItSymbolLookupResult.h [hunk 7]
```diff
@@ -1,20 +1,12 @@
...
+#import "TradeItSymbol.h"
 
 @interface TradeItSymbolLookupResult : TradeItResult
 
 // The query you passed in
 @property (nullable, copy) NSString<Optional> *query;
 
 // List of matches
-@property (nullable, copy) NSArray<Optional, TradeItSymbolLookupCompany> *results;
+@property (nullable, copy) NSArray<Optional, TradeItSymbol> *results;
```
[the remark this answers is not in the document]
Comment (jsom) [reply TradeItIosEmsApi/TradeItSymbolLookupResult.h]: Moved file.

File: TradeItIosTicketSDK2/EquitySymbolDataSource.swift (hunk 3 above)
Comment (jsom) [reply TradeItIosTicketSDK2/EquitySymbolDataSource.swift]: Yes, this PR is on hold.

File: TradeItIosTicketSDK2/EquitySymbolDataSource.swift (hunk 5 above)
Comment (jsom) [reply TradeItIosTicketSDK2/EquitySymbolDataSource.swift]: It's an array slice :<

**Comment by jsom** [comment]
@oliver-was-here most of your comments are addressed in the Yahoo PR. When I circle back to this one I'll make sure your comments are addressed.

File: TradeItIosTicketSDK2/EquitySymbolDataSource.swift (hunk 5 above)
Comment (oliver-was-here) [reply TradeItIosTicketSDK2/EquitySymbolDataSource.swift]: Ah -- so pointing to the same objects in memory? GTK.

Comment (oliver-was-here) [reply TradeItIosTicketSDK2/EquitySymbolDataSource.swift]: mm, nvm just confirmed my suspicions. This isn't doin what we think. Seems like swift array operations default to shallow copies (not sure why I was surprised by that). So, `Array(topResults)` is still going to point to the same location in memory.

Check out the swift repl and run the following commands:
```
> class A {
   var x = 5
}
> var a1 = [A(), A()]
> a1
[A] = 2 values {                                             // OUTPUT
  [0] = {
    a = 5
  }
  [1] = {
    a = 5
  }
}
> var a2 = Array(a1.prefix(1))
> a2[0].x = 6
> a1
[A] = 2 values {
  [0] = {
    a = 5
  }
  [1] = {
    a = 6
  }
}
```

Comment (jsom) [reply TradeItIosTicketSDK2/EquitySymbolDataSource.swift]: @oliver-was-here this is just converting `ArraySlice` in to an `Array`

**Comment by jsom** [comment]
Closing to launch with Y! first

**Pull request closed without merging**